### PR TITLE
[Pub/Sub] Part 2-1: New head subscription

### PR DESCRIPTION
**Pull request by czy1234**

### CODE REVIEW

File: quarkchain/cluster/jsonrpc.py [hunk 1]
```diff
@@ -258,6 +260,36 @@ def minor_block_encoder(block, include_transactions=False, extra_info=None):
     return d
 
 
+def minor_block_header_encoder(header):
+    """Encode a block header as JSON object.
+
+    :param header: a :class:`ethereum.block.Block`
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: can directly use type hinting to simplify docstring
```python
def minor_block_header_encoder(header: MinorBlockHeader) -> Dict:
...
```

File: quarkchain/cluster/jsonrpc.py [hunk 2]
```diff
@@ -1428,5 +1463,48 @@ def shutdown(self):
...
+            return {
+                "jsonrpc": "2.0",
+                "method": "eth_subscription",
+                "result": {**header_info},
+                "subscription": sub_id,
+            }
+
+        full_shard_id = shard_id_decoder(full_shard_id)
+        branch = Branch(full_shard_id)
+        shard = self.slave.shards.get(branch, None)
```
Comment (hanyunxu) [inline quarkchain/cluster/jsonrpc.py]: Maybe `shard` can be moved to `subscribe()` for the use of other future types?
`response_transcoder ()` as well.

File: quarkchain/cluster/jsonrpc.py [hunk 3]
```diff
@@ -258,6 +260,36 @@ def minor_block_encoder(block, include_transactions=False, extra_info=None):
...
+        "hashPrevRootBlock": data_encoder(header.hash_prev_root_block),
+        "nonce": quantity_encoder(header.nonce),
+        "miner": address_encoder(header.coinbase_address.serialize()),
+        "coinbase": balances_encoder(header.coinbase_amount_map),
+        "difficulty": quantity_encoder(header.difficulty),
+        "extraData": data_encoder(header.extra_data),
+        "gasLimit": quantity_encoder(header.evm_gas_limit),
+        "timestamp": quantity_encoder(header.create_time),
+    }
+    return d
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: can reuse this code for `root_block_encoder` to encode included headers and also `minor_block_encoder`

File: quarkchain/cluster/jsonrpc.py [hunk 4]
```diff
@@ -1365,6 +1397,7 @@ def get_data_default(key, decoder, default=None):
             return quantity_encoder(res) if res is not None else None
 
 
+# JSONRPC Websocket Server
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: remove

File: quarkchain/cluster/jsonrpc.py [hunk 5]
```diff
@@ -1382,7 +1415,7 @@ def __init__(
         self, env, slave_server: SlaveServer, port, host, methods: AsyncMethods
     ):
         self.loop = asyncio.get_event_loop()
-        self.port = port
+        self.port = port + int(slave_server.id[-1:])
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: should use the configured port according to #695?

File: quarkchain/cluster/jsonrpc.py [hunk 6]
```diff
@@ -1428,5 +1463,48 @@ def shutdown(self):
...
-    async def echo(self, params):
-        return params
+    async def echo(self, params, context):
+        if context is not None:
+            response = {"jsonrpc": "2.0", "result": params, "id": 1}
+            await context.send(str(response))
+        # return params
+
+    @public_methods.add
+    async def subscribe(self, sub_type, full_shard_id, context):
```
Comment (hanyunxu) [inline quarkchain/cluster/jsonrpc.py]: `qkc_subscribe`?

File: quarkchain/cluster/jsonrpc.py [hunk 7]
```diff
@@ -1394,27 +1427,29 @@ def __init__(
...
+                d = json.loads(message)
+            except Exception:
+                pass
+            method = d.get("method", "null")
+            if method in self.counters:
+                self.counters[method] += 1
+            else:
+                self.counters[method] = 1
+
+            response = await self.handlers.dispatch(message, context=websocket)
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: seems we don't need the response.

File: quarkchain/cluster/jsonrpc.py [hunk 8]
```diff
@@ -1394,27 +1427,29 @@ def __init__(
...
+            if "error" in response:
+                Logger.error(response)
+
+            # if not response.is_notification:
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: remove?

File: quarkchain/cluster/jsonrpc.py [hunk 9]
```diff
@@ -1428,5 +1463,48 @@ def shutdown(self):
         return "pong"
 
     @public_methods.add
-    async def echo(self, params):
-        return params
+    async def echo(self, params, context):
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: can remove the toy method

File: quarkchain/cluster/jsonrpc.py [hunk 10]
```diff
@@ -1428,5 +1463,48 @@ def shutdown(self):
...
+        if context is not None:
+            response = {"jsonrpc": "2.0", "result": params, "id": 1}
+            await context.send(str(response))
+        # return params
+
+    @public_methods.add
+    async def subscribe(self, sub_type, full_shard_id, context):
+        if context is None or full_shard_id is None:
+            return None
+        sub_id = os.urandom(16).hex()
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: can use context to pass the subscription ID, which increments by one

File: quarkchain/cluster/jsonrpc.py [hunk 11]
```diff
@@ -1428,5 +1463,48 @@ def shutdown(self):
...
+        response = {"jsonrpc": "2.0", "result": sub_id, "id": 1}  # TODO: change id
+        await context.send(str(response))
+        if sub_type == "newHeads":
+            await self.fetch_new_head(sub_id, full_shard_id, context)
+        else:
+            print("other types of subscription")
+            self.subscribers[sub_type].append(sub_id)
+        # return str(sub_id)
+
+    async def fetch_new_head(self, sub_id, full_shard_id, context):
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: rename `context` to `websocket`?

**Comment by ninjaahhh** [comment]
also, unit tests please

File: quarkchain/cluster/jsonrpc.py [hunk 12]
```diff
@@ -1428,5 +1420,44 @@ def shutdown(self):
         return "pong"
 
     @public_methods.add
-    async def echo(self, params):
-        return params
+    async def qkc_subscribe(self, sub_type, full_shard_id, context):
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: naming as `subscribe` should be fine

File: quarkchain/cluster/jsonrpc.py [hunk 13]
```diff
@@ -1394,27 +1388,25 @@ def __init__(
...
+        self.sub_id = 0
 
-        response = await self.handlers.dispatch(request)
-        if "error" in response:
-            Logger.error(response)
-        if not response.is_notification:
-            await websocket.send(str(response))
-        return response
+    async def __handle(self, websocket, path):
+        async for message in websocket:
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]:
```python
        sub_id = 0
        async for message in websocket:
          ...
          await self.handlers.dispatch(message, context={"websocket": websocket, "sub_id": sub_id})
          sub_id += 1
```

File: quarkchain/cluster/jsonrpc.py [hunk 14]
```diff
@@ -1428,5 +1420,44 @@ def shutdown(self):
         return "pong"
 
     @public_methods.add
-    async def echo(self, params):
-        return params
+    async def qkc_subscribe(self, sub_type, full_shard_id, context):
+        if context is None or full_shard_id is None:
+            return None
+        self.sub_id += 1
+        response = {"jsonrpc": "2.0", "result": self.sub_id, "id": 1}  # TODO: change id
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: per-connection id should be the same as in `message` in line 1395, and should be available in `d` from line 1400. can pass into `context` as well

File: quarkchain/cluster/jsonrpc.py [hunk 15]
```diff
@@ -1428,5 +1420,44 @@ def shutdown(self):
...
+        await context.send(str(response))
+
+        full_shard_id = shard_id_decoder(full_shard_id)
+        branch = Branch(full_shard_id)
+        shard = self.slave.shards.get(branch, None)
+        if not shard:
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: non-exist should close the connection. does returning none close the connection?

File: quarkchain/cluster/jsonrpc.py [hunk 16]
```diff
@@ -1428,5 +1420,44 @@ def shutdown(self):
...
+            return None
+
+        if sub_type == "newHeads":
+            await self.fetch_new_head(self.sub_id, shard, context)
+        else:
+            print("other types of subscription")
+            self.subscribers[sub_type].append(self.sub_id)
+
+    async def fetch_new_head(self, sub_id, shard, websocket):
+        latest_header = None
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: `latest_header` -> `last_header`

File: quarkchain/cluster/jsonrpc.py [hunk 17]
```diff
@@ -1428,5 +1420,44 @@ def shutdown(self):
...
+                await websocket.send(str(response))
+
+            await asyncio.sleep(0.5)
+
+    @staticmethod
+    def response_transcoder(result, sub_id):
+
+        return {
+            "jsonrpc": "2.0",
+            "method": "eth_subscription",
```
Comment (ninjaahhh) [inline quarkchain/cluster/jsonrpc.py]: 1. `eth_subscription` -> `subscription`
1. `{**result}` -> `result`

also, the format seems should be 
```
{
  "jsonrpc": xxx
  "method": yyy,
  "params": {
     "subscription": sub_id,
     "result": result
   }
}
```

File: quarkchain/cluster/tests/test_jsonrpc.py [hunk 18]
```diff
@@ -57,6 +65,46 @@ def send_request(*args):
...
+            while True:
+                response = await websocket.recv()
+                responses.append(response)
+                if len(responses) == 2:
+                    return responses
+
+    return call_async(__send_request(request))
+
+
 class TestJSONRPC(unittest.TestCase):
```
Comment (hanyunxu) [inline quarkchain/cluster/tests/test_jsonrpc.py]: Maybe we can change `TestJSONRPC` to `TestJSONRPCHttp` and create another class `TestJSONRPCWebsocket` specifically for jsonrpc-websocket test.

File: quarkchain/cluster/tests/test_jsonrpc.py [hunk 19]
```diff
@@ -1131,3 +1190,45 @@ def test_createTransactions(self):
...
+        acc1 = Address.create_from_identity(id1, full_shard_key=0)
+
+        with ClusterContext(
+            1, acc1, small_coinbase=True
+        ) as clusters, jrpc_server_context(
+            clusters[0].master
+        ), jrpc_websocket_server_context(
+            clusters[0].slave_list[0]
+        ):
+            # master = clusters[0].master
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_jsonrpc.py]: remove lone 1205 to line 1211

File: quarkchain/cluster/tests/test_jsonrpc.py [hunk 20]
```diff
@@ -57,6 +65,57 @@ def send_request(*args):
...
+    responses = []
+
+    async def __send_request(request):
+        uri = "ws://0.0.0.0:38591"
+        async with websockets.connect(uri) as websocket:
+            await websocket.send(request)
+            while True:
+                response = await websocket.recv()
+                responses.append(response)
+                if len(responses) == 2:
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_jsonrpc.py]: 2 can be passed in as an argument in `send_websocket_request` with a default value (like 1)

File: quarkchain/cluster/tests/test_jsonrpc.py [hunk 21]
```diff
@@ -57,6 +65,57 @@ def send_request(*args):
...
+    env.cluster_config.SLAVE_LIST.append(slave_config)
+    env.slave_config = env.cluster_config.get_slave_config("S1")
+    server = JSONRPCWebsocketServer.start_websocket_server(env, slave_server)
+    try:
+        yield server
+    finally:
+        server.shutdown()
+
+
+def send_websocket_request(request):
```
Comment (ninjaahhh) [inline quarkchain/cluster/tests/test_jsonrpc.py]: maybe this?
https://jsonrpcclient.readthedocs.io/en/latest/examples.html#websockets

image

**Pull request merged**